system/security/plugins/jwtSecurity/jwtEndpoint.cpp

```diff
@@ -0,0 +1,257 @@
+        credentialsJSON["client_id"] = clientID;
+         if (!nonce.empty())
+            credentialsJSON["nonce"] = nonce;
+        credentialsStr = credentialsJSON.dump();
+
+        try
+        {
+            // Initialize our capture buffer to a reasonable size to avoid
+            // memory reallocation
+            InitMemoryStructCapacity(captureBuffer, INITIAL_BUFFER_SIZE);
+
+            headers = curl_slist_append(headers, "Content-Type: application/json;charset=UTF-8");
+            curl_easy_setopt(curlHandle, CURLOPT_HTTPHEADER, headers);
+
+            curl_easy_setopt(curlHandle, CURLOPT_URL, jwtEndPoint.c_str());
+            curl_easy_setopt(curlHandle, CURLOPT_POST, 1);
+            curl_easy_setopt(curlHandle, CURLOPT_POSTFIELDS, credentialsStr.c_str());
+            if (allowSelfSignedCert)
+                curl_easy_setopt(curlHandle, CURLOPT_SSL_VERIFYPEER, 0);
+            curl_easy_setopt(curlHandle, CURLOPT_FOLLOWLOCATION, 1);
+            curl_easy_setopt(curlHandle, CURLOPT_UNRESTRICTED_AUTH, 1);
+            curl_easy_setopt(curlHandle, CURLOPT_NOPROGRESS, 1);
+            curl_easy_setopt(curlHandle, CURLOPT_WRITEFUNCTION, CaptureIncomingCURLReply);
+            curl_easy_setopt(curlHandle, CURLOPT_WRITEDATA, static_cast<void*>(&captureBuffer));
+            curl_easy_setopt(curlHandle, CURLOPT_ERRORBUFFER, curlErrBuffer);
+
+            curlResponseCode = curl_easy_perform(curlHandle);
+
+            if (curlResponseCode == CURLE_OK && captureBuffer.size > 0)
+            {
+                std::string     responseStr = std::string(captureBuffer.memory, captureBuffer.size);
+
+                // Poor check to see if reply resembles JSON
+                if (responseStr[0] == '{')
+                    apiResponse = responseStr;
+                else
+                    PROGLOG("CJwtSecurityManager: Invalid JWT endpoint response: %s", responseStr.c_str());
+            }
+            else
+            {
+                if (curlResponseCode != CURLE_OK)
+                    PROGLOG("CJwtSecurityManager login error: libcurl error (%d): %s", curlResponseCode, (curlErrBuffer[0] ? curlErrBuffer : "<unknown>"));
+                else
+                    PROGLOG("CJwtSecurityManager login error: No content from JWT endpoint");
+            }
+        }
+        catch (...)
+        {
+            // We should not be allowing exceptions to propagate up
+            // the chain; ignore the error (and let's hope the source
+            // of the error logged the details)
+        }
+
+        // Cleanup
+        if (headers)
+        {
+            curl_slist_free_all(headers);
+            headers = nullptr;
+        }
+        curl_easy_cleanup(curlHandle);
+        FreeMemoryStruct(captureBuffer);
+    }
+
+    return apiResponse;
+}
+
+std::string tokenFromRefresh(const std::string& jwtEndPoint, bool allowSelfSignedCert, const std::string& clientID, const std::string& refreshToken)
```

Comment:
With the exception of the json parameters, and a couple logging statements, this is identical to tokenFromLogin. Could they be refactored to share code instead of replicating?